src/audio/module_adapter/module_adapter_ipc3.c

```diff
@@ -207,7 +207,7 @@ static int module_adapter_get_set_params(struct comp_dev *dev, struct sof_ipc_ct
 
 	comp_warn(dev, "module_adapter_get_set_params(): no configuration op set for %d",
 		  dev_comp_id(dev));
-	return 0;
+	return -EIO; /* non-implemented error */
```

Comment:
I realized this leads to a special case that by calling this -EIO will be returned for CMD_SET_DATA if `set_configuration` is not implemented. It's beyond my purpose that returning -EIO for CMD_GET_DATA only.

Hi @LaurentiuM1234 , I just saw your comment saying that `cadence.c` doesn't implement both `get_configuration` and `set_configuration`, is that correct? If the non-implemented `set_configuration` is intentional, I think this line should be fixed to keep CMD_SET_DATA behavior as is.

Reply:
> I realized this leads to a special case that by calling this -EIO will be returned for CMD_SET_DATA if `set_configuration` is not implemented. It's beyond my purpose that returning -EIO for CMD_GET_DATA only.

ACK, so now `get_configuration()` becomes mandatory and `set_configuration()` remains optional. Might be out of scope for this, but did you consider adding all the `interface->set/get_configuration()` logic into some wrapper functions?

i.e: something like:

```
static inline int module_set_configuration(struct processing_module *mod, uint32_t config_id
                                                                   enum module_cfg_fragment_position pos, uint32_t data_offset_size,
                                                                   const uint8_t *fragment, size_t fragment_size, uint8_t *response,
                                                                   size_t response_size)
{
    const struct module_interface *const ops = mod->dev->drv->adapter_ops;

    if (ops->set_configuration) {
       // call set_configuration
    }

    // optional operation
    return 0;
}

static inline int module_get_configuration(struct processing_module *mod, uint32_t config_id
                                                                   enum module_cfg_fragment_position pos, uint32_t data_offset_size,
                                                                   const uint8_t *fragment, size_t fragment_size, uint8_t *response,
                                                                   size_t response_size)
{
    const struct module_interface *const ops = mod->dev->drv->adapter_ops;

    // this is not optional
    if (!ops->get_configuration)
        return -EIO;

    // some more code here
}
```

Since this is also used by the IPC4 stuff, having a centralized view showing which operation is optional and which isn't might be helpful?


> Hi @LaurentiuM1234 , I just saw your comment saying that `cadence.c` doesn't implement both `get_configuration` and `set_configuration`, is that correct? If the non-implemented `set_configuration` is intentional, I think this line should be fixed to keep CMD_SET_DATA behavior as is.

The `cadence` module doesn't implement the `get_configuration()` stuff so if we make it mandatory the issue will still persist. Can't really tell you why it's not implemented but we can probably get around the issue by just implementing a dummy `get_configuration()` if we really don't need a proper one. Anyways, IMO not a blocker for this.

Reply:
Thanks. Some contexts below for clearer statement:

The commit doesn't mean to make `get_configuration` mandatory (in a literal manner). Instead, the intention is to fix the use case of SOF build which integrates the instance without `get_configuration` in module_adapter.

Consider a host running the IPC command to read blob data from the module_adapter bytes control, if module_adapter doesn't return error while the instance doesn't implement `get_configuration`, the host will consider the blob data is read back successfully, leading to problems on wrong (or illegal) data blob access. As a consequence, it should be better to return error than zero, provided that the host has no clue whether `get_configuration` is implemented by instance or not.